pylint/checkers/refactoring.py

```diff
@@ -216,10 +222,39 @@ def visit_ifexp(self, _):
     def visit_comprehension(self, node):
         self._if_counter += len(node.ifs)
 
-    @utils.check_messages('too-many-nested-blocks', 'simplifiable-if-statement')
+    def _always_return(self, body):
+        """Search if a list of node has 'return', search nested return into ifs
+        """
+        alwr_all = []
+        for item in body:
+            if isinstance(item, astroid.Return):
+                return True
+            elif isinstance(item, astroid.If):
+                if not hasattr(item, 'alwr_if'):
+                    # Assign variable to object in order to re-use computed data
+                    item.alwr_if = self._always_return(item.body)
+                    item.alwr_else = self._always_return(item.orelse) \
+                        if item.orelse else False
+                alwr_all.append(item.alwr_if and item.alwr_else)
+        return all(alwr_all) if alwr_all else False
+
+    def _check_superfluous_else_return(self, node):
+        if not node.orelse:
+            # Not interested in if statements without else.
+            return
+        if not hasattr(node, 'alwr_if'):
```

Comment:
No need for this. Check the snippet of code from the top, this could be used in instead of what we are having in this PR. As a caching mechanism, if you need it, you can annotate the if specialization function with functools.lru_cache.